File: O2Core/DeltaCompress.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.PortableExecutable;
using System.Text;
using System.Threading.Tasks;

namespace Oxygen
{
    /// <summary>
    /// Performs compression by calculating the difference between two sets of data and then
    /// performing run length encoding.
    /// </summary>
    /// <example>
    /// <code>
    /// byte[] t1 = new byte[] { 0, 1, 3, 4, 5, 1, 4, };
    /// byte[] t2 = new byte[] { 0, 1, 3, 6, 5, 1, 4, };
    ///
    /// byte[] result = DeltaCompress.Compress(t1, t2);
    ///
    /// byte[] o = DeltaCompress.Decompress(t1, result);
    /// </code>
    /// </example>
    public static class DeltaCompress
    {
        private const byte UNCOMPRESSED_BLOCK = 0;
        private const byte DELTA_COMPRESSED_BLOCK = 1;

        private static byte[] CalculateDeltas(byte[] initialData, byte[] newData, int length, int initialDataOffset, int newDataOffset)
        {
            byte[] delta = new byte[length];
            for (int i = 0; i < newData.Length; i++)
            {
                delta[i] = (byte)(initialData[i + initialDataOffset] - newData[i + newDataOffset]);
            }

            return delta;
        }

        private static void WriteBlock(MemoryStream ms, byte[] data, byte flags, int offset, int length)
        {
            ms.WriteByte(flags);

            if (flags == DELTA_COMPRESSED_BLOCK)
            {
                ms.WriteByte((byte)(data.Length & 0xFF));
                ms.WriteByte((byte)((data.Length >> 8) & 0xFF));

                // Then run length encode the data.
                byte value = data[0];
                short count = 1;
                for (int i = 1; i < data.Length; i++)
                {
                    if (data[i] == value)
                    {
                        count++;
                    }
                    else
                    {
                        ms.WriteByte(value);
                        ms.WriteByte((byte)(count & 0xFF));
                        ms.WriteByte((byte)((count >> 8) & 0xFF));

                        value = data[i];
                        count = 1;
                    }
                }

                ms.WriteByte(value);
                ms.WriteByte((byte)(count & 0xFF));
                ms.WriteByte((byte)((count >> 8) & 0xFF));
            }
            else if (flags == UNCOMPRESSED_BLOCK)
            {
                ms.WriteByte((byte)(length & 0xFF));
                ms.WriteByte((byte)((length >> 8) & 0xFF));
                ms.Write(data, offset, length);
            }
        }

        public static byte[] Compress(byte[] initialData, byte[] newData)
        {
            using (MemoryStream ms = new MemoryStream())
            {
                if (initialData.Length > newData.Length)
                {
                    //int offset = FindSubArray(initialData, newData);
                    //if (offset > -1)
                    //{
                    //    WriteData(ms, initialData, newData, offset);
                    //}
                    //else
                    {
                        WriteBlock(ms, newData, UNCOMPRESSED_BLOCK, 0, newData.Length);
                    }
                }
                else if (newData.Length > initialData.Length)
                {
                    int offset = FindSubArray(newData, initialData);
                    if (offset > -1)
                    {
                        WriteData(ms, newData, initialData, offset);
                    }
                    else
                    {
                        WriteBlock(ms, newData, UNCOMPRESSED_BLOCK, 0, newData.Length);
                    }
                }
                else
                {
                    byte[] delta;
                    delta = CalculateDeltas(initialData, newData, initialData.Length, 0, 0);
                    WriteBlock(ms, delta, DELTA_COMPRESSED_BLOCK, 0, initialData.Length);
                }

                return ms.ToArray();
            }
        }

        private static void WriteData(MemoryStream ms, byte[] longer, byte[] shorter, int offset)
        {
            int length = shorter.Length;
            byte[] delta = CalculateDeltas(longer, shorter, length, offset, 0);

            if (offset > 0)
            {
                int uncompressedLength = offset;
                WriteBlock(ms, longer, UNCOMPRESSED_BLOCK, 0, uncompressedLength);
            }

            WriteBlock(ms, delta, DELTA_COMPRESSED_BLOCK, offset, length);

            if (offset + length < longer.Length)
            {
                int uncompressedLength = offset + length;
                WriteBlock(ms, longer, UNCOMPRESSED_BLOCK, uncompressedLength, longer.Length - uncompressedLength);
            }
        }

        private static void ReadBlock(MemoryStream ms, List<byte> newData, byte[] initialData, byte[] delta, ref int pos)
        {
            int flags = delta[pos++];
            if (flags == UNCOMPRESSED_BLOCK)
            {
                int countLo = delta[pos++];
                int countHi = delta[pos++];

                int count = countLo | (countHi << 8);

                for (int i = 0; i < count; i++)
                {
                    newData.Add(delta[pos++]);
                }
            }
            else if (flags == DELTA_COMPRESSED_BLOCK)
            {
                int dataLength;
                {
                    int countLo = delta[pos++];
                    int countHi = delta[pos++];

                    dataLength = countLo | (countHi << 8);
                }

                // Decode the run length encoding.
                int offset = newData.Count;
                int end = newData.Count + dataLength;
                while (newData.Count < end)
                {
                    int value = delta[pos++];
                    int countLo = delta[pos++];
                    int countHi = delta[pos++];

                    int count = countLo | (countHi << 8);

                    for (int j = 0; j < count; j++)
                    {
                        newData.Add((byte)value);
                    }
                }

                // Decode the deltas.
                for (int i = 0; i < initialData.Length; i++)
                {
                    newData[i + offset] = (byte)(initialData[i] - newData[i + offset]);
                }
            }
        }

        public static byte[] Decompress(byte[] initialData, byte[] delta)
        {
            List<byte> newData = new List<byte>();
            using (MemoryStream ms = new MemoryStream())
            {
                int pos = 0;
                while (pos < delta.Length)
                {
                    ReadBlock(ms, newData, initialData, delta, ref pos);
                }
            }

            return newData.ToArray();
        }

        /// <summary>
        /// Searches for the shorter array in the longer array.
        /// </summary>
        /// <param name="longer">The longer array.</param>
        /// <param name="shorter">The shorter array.</param>
        /// <exception cref="ArgumentException">Throws if the parameters are invalid.</exception>
        /// <returns>Returns the index of where the sub array is located, or -1 if the sub array could not be found.</returns>
        private static int FindSubArray(byte[] longer, byte[] shorter)
        {
            if (longer.Length <= shorter.Length)
            {
                throw new ArgumentException($"Invalid arguments '{nameof(longer)}' should be a longer array than '{nameof(shorter)}'.");
            }

            for (int i = 0; i < longer.Length; i++)
            {
                bool success = true;
                for (int j = 0; j < shorter.Length; j++)
                {
                    if (longer[i + j] != shorter[j])
                    {
                        success = false;
                        break;
                    }
                }

                if (success)
                {
                    return i;
                }
            }

            return -1;
        }
    }
}

File: libOxygen/DeltaCompress.cpp
#include "DeltaCompress.h"

using namespace Oxygen;

constexpr char UNCOMPRESSED_BLOCK = 0;
constexpr char DELTA_COMPRESSED_BLOCK = 1;

static void CalculateDeltas(
    const unsigned char* initialData,
    const unsigned char* newData,
    int initialDataOffset, 
    int newDataOffset,
    int newDataLength,
    std::vector<unsigned char> &delta)
{
    for (int i = 0; i < newDataLength; i++)
    {
        delta.push_back((unsigned char)(initialData[i + initialDataOffset] - newData[i + newDataOffset]));
    }
}

static void WriteBlock(
    std::vector<unsigned char>& stream,
    const unsigned char* data,
    int dataLength,
    char flags,
    int offset)
{
    stream.push_back(flags);

    if (flags == DELTA_COMPRESSED_BLOCK)
    {
        stream.push_back((unsigned char)(dataLength & 0xFF));
        stream.push_back((unsigned char)((dataLength >> 8) & 0xFF));

        // Then run length encode the data.
        unsigned char value = data[0];
        int count = 1;
        for (int i = 1; i < dataLength; i++)
        {
            if (data[i] == value)
            {
                count++;
            }
            else
            {
                stream.push_back(value);
                stream.push_back((unsigned char)(count & 0xFF));
                stream.push_back((unsigned char)((count >> 8) & 0xFF));

                value = data[i];
                count = 1;
            }
        }

        stream.push_back(value);
        stream.push_back((unsigned char)(count & 0xFF));
        stream.push_back((unsigned char)((count >> 8) & 0xFF));
    }
    else if (flags == UNCOMPRESSED_BLOCK)
    {
        stream.push_back((unsigned char)(dataLength & 0xFF));
        stream.push_back((unsigned char)((dataLength >> 8) & 0xFF));

        for (int i = 0; i < dataLength; i++)
        {
            stream.push_back((unsigned char)data[offset + i]);
        }
    }
}

static int FindSubArray(const unsigned char* longer, const unsigned char* shorter, int longerLength, int shorterLength)
{
    if (longerLength <= shorterLength)
    {
        return -2;
    }

    for (int i = 0; i < longerLength; i++)
    {
        bool success = true;
        for (int j = 0; j < shorterLength; j++)
        {
            if (longer[i + j] != shorter[j])
            {
                success = false;
                break;
            }
        }

        if (success)
        {
            return i;
        }
    }

    return -1;
}

static void WriteData(std::vector<unsigned char>& stream,
    const unsigned char* longer, const unsigned char* shorter,
    int longerLength, int shorterLength, int offset)
{
    int length = shorterLength;
    std::vector<unsigned char> delta;
    CalculateDeltas(longer, shorter, length, offset, 0, delta);

    if (offset > 0)
    {
        const int uncompressedLength = offset;
        WriteBlock(stream, longer, UNCOMPRESSED_BLOCK, 0, uncompressedLength);
    }

    WriteBlock(stream, delta.data(), int(delta.size()), DELTA_COMPRESSED_BLOCK, offset);

    if (offset + length < longerLength)
    {
        const int uncompressedLength = offset + length;
        WriteBlock(stream, longer, UNCOMPRESSED_BLOCK, uncompressedLength, longerLength - uncompressedLength);
    }
}

int Oxygen::Compress(const unsigned char* initialData, int numInitialBytes, const unsigned char* newData, int numNewDataBytes, unsigned char** deltaData)
{
    std::vector<unsigned char> stream;
    std::vector<unsigned char> delta;

    if (numInitialBytes > numNewDataBytes)
    {
        WriteBlock(stream, newData, numNewDataBytes, UNCOMPRESSED_BLOCK, 0);
    }
    else if (numNewDataBytes > numInitialBytes)
    {
        int offset = FindSubArray(newData, initialData, numNewDataBytes, numInitialBytes);
        if (offset > -1)
        {
            WriteData(stream, newData, initialData, numNewDataBytes, numInitialBytes, offset);
        }
        else
        {
            WriteBlock(stream, newData, numNewDataBytes, UNCOMPRESSED_BLOCK, 0);
        }
    }
    else
    {
        CalculateDeltas(initialData, newData, 0, 0, numNewDataBytes, delta);
        WriteBlock(stream, delta.data(), int(delta.size()), DELTA_COMPRESSED_BLOCK, 0);
    }

    *deltaData = new unsigned char[stream.size()];
    std::memcpy(*deltaData, stream.data(), stream.size());

    return int(stream.size());
}

static void ReadBlock(
    std::vector<unsigned char>& newData,
    const unsigned char* initialData,
    const unsigned char* delta,
    int numInitialData,
    int& pos)
{
    int flags = delta[pos++];
    if (flags == UNCOMPRESSED_BLOCK)
    {
        int countLo = delta[pos++];
        int countHi = delta[pos++];

        int count = countLo | (countHi << 8);

        for (int i = 0; i < count; i++)
        {
            newData.push_back(delta[pos++]);
        }
    }
    else if (flags == DELTA_COMPRESSED_BLOCK)
    {
        int dataLength;
        {
            int countLo = delta[pos++];
            int countHi = delta[pos++];

            dataLength = countLo | (countHi << 8);
        }

        // Decode the run length encoding.
        size_t offset = newData.size();
        int end = int(newData.size()) + dataLength;
        while (newData.size() < end)
        {
            int value = delta[pos++];
            int countLo = delta[pos++];
            int countHi = delta[pos++];

            int count = countLo | (countHi << 8);

            for (int j = 0; j < count; j++)
            {
                newData.push_back((char)value);
            }
        }

        // Decode the deltas.
        for (int i = 0; i < numInitialData; i++)
        {
            newData[i + offset] = (char)(initialData[i] - newData[i + offset]);
        }
    }
}

void Oxygen::Decompress(unsigned char* initialData, int numInitialBytes, unsigned char* delta, int numDeltaBytes, std::vector<unsigned char>& newData)
{
    int pos = 0;
    while (pos < numDeltaBytes)
    {
        ReadBlock(newData, initialData, delta, numInitialBytes, pos);
    }
}
